common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -6393,106 +6395,64 @@ static void NativeCrypto_SSL_set1_tls_channel_id(JNIEnv* env, jclass,
-    if (ssl == nullptr) {
-        return;
-    }
-
-    if (certificatesJava == nullptr) {
+    if (encodedCertificatesJava == nullptr) {
         Errors::jniThrowNullPointerException(env, "certificates == null");
-        JNI_TRACE("ssl=%p NativeCrypto_SSL_use_certificate => certificates == null", ssl);
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_set_chain_and_key => certificates == null", ssl);
         return;
     }
-
-    size_t length = static_cast<size_t>(env->GetArrayLength(certificatesJava));
-    if (length == 0) {
-        Errors::jniThrowException(env, "java/lang/IllegalArgumentException", "certificates.length == 0");
-        JNI_TRACE("ssl=%p NativeCrypto_SSL_use_certificate => certificates.length == 0", ssl);
+    size_t numCerts = static_cast<size_t>(env->GetArrayLength(encodedCertificatesJava));
+    if (numCerts == 0) {
+        Errors::jniThrowException(env, "java/lang/IllegalArgumentException",
+                                  "certificates.length == 0");
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_set_chain_and_key => certificates.length == 0", ssl);
         return;
     }
-
-    ScopedLongArrayRO certificates(env, certificatesJava);
-    if (certificates.get() == nullptr) {
-        JNI_TRACE("ssl=%p NativeCrypto_SSL_use_certificate => certificates == null", ssl);
+    if (pkeyRef == nullptr) {
+        Errors::jniThrowNullPointerException(env, "privateKey == null");
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_set_chain_and_key => privateKey == null", ssl);
         return;
     }
 
-    X509* serverCert = reinterpret_cast<X509*>(static_cast<uintptr_t>(certificates[0]));
-    if (serverCert == nullptr) {
-        // Note this shouldn't happen since we checked the number of certificates above.
-        Errors::jniThrowOutOfMemory(env, "Unable to allocate local certificate chain");
-        JNI_TRACE("ssl=%p NativeCrypto_SSL_use_certificate => chain allocation error", ssl);
+    // Get the private key.
+    EVP_PKEY* pkey = fromContextObject<EVP_PKEY>(env, pkeyRef);
+    if (pkey == nullptr) {
+        Errors::jniThrowNullPointerException(env, "pkey == null");
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_set_chain_and_key => pkey == null", ssl);
         return;
     }
 
-    int ret = SSL_use_certificate(ssl, serverCert);
-    if (ret != 1) {
-        ALOGE("%s", ERR_error_string(ERR_peek_error(), nullptr));
-        Errors::throwSSLExceptionWithSslErrors(env, ssl, SSL_ERROR_NONE, "Error setting certificate");
-        safeSslClear(ssl);
-        JNI_TRACE("ssl=%p NativeCrypto_SSL_use_certificate => SSL_use_certificate error", ssl);
-        return;
-    }
+    // Copy the certificates.
+    std::vector<bssl::UniquePtr<CRYPTO_BUFFER>> certBufferRefs(numCerts);
+    std::vector<CRYPTO_BUFFER*> certBuffers(numCerts);
+    for(size_t i = 0; i < numCerts; ++i) {
+        const jbyteArray encodedCertJava =
+            (jbyteArray) env->GetObjectArrayElement(encodedCertificatesJava, i);
```

Comment:
Is this a leak? Every other use of `GetObjectArrayElement` puts it in a `ScopedLocalRef`.